src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/KeyValuePairConverter.cs

```diff
@@ -91,28 +88,16 @@ internal override void Initialize(JsonSerializerOptions options, JsonTypeInfo? j
             return true;
         }
 
-        protected override void EndRead(ref ReadStack state)
-        {
-            Debug.Assert(state.Current.PropertyIndex == 0);
-
-            if (state.Current.CtorArgumentState!.ParameterIndex != NumProperties)
-            {
-                ThrowHelper.ThrowJsonException();
-            }
-        }
-
         private bool FoundKeyProperty(string propertyName, bool caseInsensitiveMatch)
         {
             return propertyName == _keyName ||
-                (caseInsensitiveMatch && string.Equals(propertyName, _keyName, StringComparison.OrdinalIgnoreCase)) ||
-                propertyName == KeyNameCLR;
+                (caseInsensitiveMatch && string.Equals(propertyName, _keyName, StringComparison.OrdinalIgnoreCase));
```

Comment:
This predates your own changes, but doesn't this make two string comparisons in the case insensitive case? IOW shouldn't this be
```suggestion
                return caseInsensitiveMatch ? string.Equals(propertyName, _keyName, StringComparison.OrdinalIgnoreCase) : propertyName == _valueName;
```